components/function-controller/internal/resource/resource.go

```diff
@@ -0,0 +1,72 @@
+package resource
+
+import (
+	"context"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apilabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type Resource interface {
+	Create(context.Context, Object, Object) error
+	ListByLabel(context.Context, string, map[string]string, runtime.Object) error
+	DeleteAllByLabel(context.Context, Object, string, map[string]string) error
+}
+
+//go:generate mockery -name=Client -output=automock -outpkg=automock -case=underscore
+type Client interface {
+	Create(context.Context, runtime.Object, ...client.CreateOption) error
+	List(context.Context, runtime.Object, ...client.ListOption) error
+	DeleteAllOf(context.Context, runtime.Object, ...client.DeleteAllOfOption) error
+}
+
+type Object interface {
+	runtime.Object
+	metav1.Object
+}
+
+var _ Resource = &resourceSvc{}
+
+type resourceSvc struct {
+	client Client
+	schema *runtime.Scheme
+}
+
+func New(client Client, schema *runtime.Scheme) *resourceSvc {
+	return &resourceSvc{
+		client: client,
+		schema: schema,
+	}
+}
+
+func (r *resourceSvc) Create(ctx context.Context, parent, object Object) error {
+	if err := controllerutil.SetControllerReference(parent, object, r.schema); err != nil {
+		return err
+	}
+
+	return r.client.Create(ctx, object)
+}
+
+func (r *resourceSvc) ListByLabel(ctx context.Context, namespace string, labels map[string]string, list runtime.Object) error {
```

Comment:
is there a way to compile time check whether `list runtime.Object` is really a list? I can see that in unstructured interface there's a method `isList`, but it would be then necessary to change type - what do you think about it? 

disclaimer: it's a nitpick, ignore if it's too much work for now

Reply:
I would leave it as it is currently